src/main/scala/org/apache/spark/shuffle/ucx/UcxShuffleTransport.scala

```diff
@@ -0,0 +1,229 @@
+   * This depends on [[ ShuffleTransport.progress() ]] calls,
+   * and does not indicate actual data transfer time.
+   */
+  override def getElapsedTimeNs: Long = endTime - startTime
+
+  /**
+   * Indicates number of valid bytes in receive memory when using
+   * [[ ShuffleTransport.fetchBlockByBlockId()]]
+   */
+  override def recvSize: Long = receiveSize
+}
+
+class UcxRequest(request: UcpRequest, stats: OperationStats) extends Request {
+
+  override def isCompleted: Boolean = request.isCompleted
+
+  override def cancel(): Unit = request.close()
+
+  override def getStats: Option[OperationStats] = Some(stats)
+}
+
+/**
+ * UCX implementation of [[ ShuffleTransport ]] API
+ */
+class UcxShuffleTransport(var ucxShuffleConf: UcxShuffleConf = null)
+  extends ShuffleTransport with Logging {
+  // UCX entities
+  private var ucxContext: UcpContext = _
+  private var globalWorker: UcpWorker = _
+  private val ucpWorkerParams = new UcpWorkerParams()
+
+  // TODO: reimplement as workerPool, since spark may create/destroy threads dynamically
+  private var threadLocalWorker: ThreadLocal[UcxWorkerWrapper] = _
+  private val allocatedWorkers = ConcurrentHashMap.newKeySet[UcxWorkerWrapper]
+  private var progressThread: Thread = _
+
+  private val registeredBlocks: ConcurrentHashMap[BlockId, Block] = new ConcurrentHashMap
+  private val memMapParams = new UcpMemMapParams()
+
+  if (ucxShuffleConf.useOdp) {
+    memMapParams.nonBlocking()
+  }
+
+  //Mapping between executorId and it's address
+  private[ucx] val executorIdToAddress = new ConcurrentHashMap[String, ByteBuffer]()
+  private[ucx] val clientConnections = new ConcurrentHashMap[ByteBuffer, UcpEndpoint]()
+
+  var memoryPool: MemoryPool = _
+  /**
+   * Initialize transport resources. This function should get called after ensuring that SparkConf
+   * has the correct configurations since it will use the spark configuration to configure itself.
+   */
+  override def init(): ByteBuffer = {
+    if (ucxShuffleConf == null) {
+      ucxShuffleConf = new UcxShuffleConf(SparkEnv.get.conf)
+    }
+
+    val params = new UcpParams().requestTagFeature().requestWakeupFeature()
+    if (ucxShuffleConf.protocol == "one-sided") {
+      params.requestRmaFeature()
+    }
+    ucxContext = new UcpContext(params)
+    globalWorker = ucxContext.newWorker(new UcpWorkerParams().requestWakeupTagRecv().requestWakeupRX())
+
+    val result = globalWorker.getAddress
+    require(result.capacity <= ucxShuffleConf.maxWorkerAddressSize,
+      s"${ucxShuffleConf.WORKER_ADDRESS_SIZE.key} < ${result.capacity}")
+
+    memoryPool = new UcxHostBounceBuffersPool(ucxShuffleConf, ucxContext)
+    progressThread = new GlobalWorkerRpcThread(globalWorker, memoryPool, this)
+
+    threadLocalWorker = ThreadLocal.withInitial(() => {
+      val localWorker = ucxContext.newWorker(ucpWorkerParams)
+      val result = new UcxWorkerWrapper(localWorker, this, ucxShuffleConf, memoryPool)
```

Comment:
we use result here, and also in line 109 for the ByteBuffer. Nit mostly, but we can change this one to `workerWrapper`?